expo/eas-cli
# [eas-cli] Add support for Robot users

# Why

This adds support for robot users, through `EXPO_TOKEN=<robot-user-token> eas ...`. 

# Future stuff

We could wait for the outcome of the `Actor.displayName` discussion. That could replace the `getActorUserName` helper and make our usernames easier to use.

# How

It's similar to [PR #2440](https://github.com/expo/expo-cli/pull/2440), but implemented in such way that it's heavily reliant on the `CurrentUserQuery` result. If we add properties, like `isExpoAdmin` to this query, they are propagated everywhere once you run `$ yarn generate-graphql-code`.

This has a lot of benefits, including easier to maintain and more type-related to the output of the API. But, relating it more the GraphQL API also has its downsides. For example, the `meActor.accounts` is typed in GraphQL as `accounts: [Account]` meaning:
- `meActor.accounts` can be `null`
- `meActor.accounts` can be an `array`
  - this array can contain `null`
  - this array can contain an object with `id` and `name`

This type of arrays are pure suffering in TypeScript, because we have to account for the possible "`null` or `array`" _AND_ "`null` within the `array`". Would be nicer to have this typed as `accounts: [Account!]!` where possible, it still may contain an empty array but it's way easier to not-have-to-account-for all those possible `null` values.

# Test Plan

- Updated all tests, run `yarn test`
- Add additional tests to both `project` and `user`

**To test manually**
- Create a robot user for your account (see GraphQL query below, not yet available on website)
- Create an access token for your robot user ([using this link](https://staging.expo.io/settings/access-tokens))
- Try smoke testing commands with `EXPO_TOKEN=<robot-token> eas ...` (e.g. `eas whoami`)

<details>
<summary>Create a robot with graphql mutation</summary>
<pre>
mutation {
  robot {
    createRobotForAccount(
      accountID: "{id}"
      robotData: { name: "GLaDOS" }
      permissions: PUBLISH
    ) {
      id
      firstName
    }
  }
}
</pre>
</details>
